File: alligator/src/main/java/me/aartikov/alligator/converters/DefaultFragmentConverter.java
package me.aartikov.alligator.converters;

import java.io.Serializable;

import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.app.Fragment;

import me.aartikov.alligator.Screen;

/**
 * Date: 15.09.2018
 * Time: 10:18
 *
 * @author Artur Artikov
 */

/**
 * Creates a fragment of the given class. It also puts a screen to the fragment's arguments if {@code ScreenT} is {@code Serializable} or {@code Parcelable}.
 *
 * @param <ScreenT> screen type
 */

public class DefaultFragmentConverter<ScreenT extends Screen> implements FragmentConverter<ScreenT> {
	private static final String KEY_SCREEN = "me.aartikov.alligator.KEY_SCREEN";

	private Class<ScreenT> mScreenClass;
	private Class<? extends Fragment> mFragmentClass;

	public DefaultFragmentConverter(Class<ScreenT> screenClass, Class<? extends Fragment> fragmentClass) {
		mScreenClass = screenClass;
		mFragmentClass = fragmentClass;
	}

	@Override
	public Fragment createFragment(ScreenT screen) {
		try {
			Fragment fragment = mFragmentClass.newInstance();
			if (screen instanceof Serializable) {
				Bundle arguments = new Bundle();
				arguments.putSerializable(KEY_SCREEN, (Serializable) screen);
				fragment.setArguments(arguments);
			} else if (screen instanceof Parcelable) {
				Bundle arguments = new Bundle();
				arguments.putParcelable(KEY_SCREEN, (Parcelable) screen);
				fragment.setArguments(arguments);
			}
			return fragment;
		} catch (InstantiationException e) {
			throw new RuntimeException("Failed to create a fragment", e);
		} catch (IllegalAccessException e) {
			throw new RuntimeException("Failed to create a fragment", e);
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public ScreenT getScreen(Fragment fragment) {
		if (fragment.getArguments() == null) {
			throw new IllegalArgumentException("Fragment has no arguments.");
		} else if (Serializable.class.isAssignableFrom(mScreenClass)) {
			return (ScreenT) fragment.getArguments().getSerializable(KEY_SCREEN);
		} else if (Parcelable.class.isAssignableFrom(mScreenClass)) {
			return (ScreenT) fragment.getArguments().getParcelable(KEY_SCREEN);
		} else {
			throw new IllegalArgumentException("Screen " + mScreenClass.getSimpleName() + " should be Serializable or Parcelable.");
		}
	}
}

File: alligator/src/main/java/me/aartikov/alligator/converters/DefaultFragmentConverter.kt
package me.aartikov.alligator.converters

import android.os.Bundle
import android.os.Parcelable
import androidx.fragment.app.Fragment
import me.aartikov.alligator.Screen
import java.io.Serializable

/**
 * Creates a fragment of the given class. It also puts a screen to the fragment's arguments if `ScreenT` is `Serializable` or `Parcelable`.
 *
 * @param <ScreenT> screen type
</ScreenT> */
class DefaultFragmentConverter<ScreenT : Screen>(
    private val mScreenClass: Class<ScreenT>,
    private val mFragmentClass: Class<out Fragment>
) : FragmentConverter<ScreenT> {
    override fun createFragment(screen: ScreenT): Fragment {
        return try {
            val fragment = mFragmentClass.newInstance()
            if (screen is Serializable) {
                val arguments = Bundle()
                arguments.putSerializable(KEY_SCREEN, screen as Serializable)
                fragment.arguments = arguments
            } else if (screen is Parcelable) {
                val arguments = Bundle()
                arguments.putParcelable(KEY_SCREEN, screen as Parcelable)
                fragment.arguments = arguments
            }
            fragment
        } catch (e: InstantiationException) {
            throw RuntimeException("Failed to create a fragment", e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Failed to create a fragment", e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun getScreen(fragment: Fragment): ScreenT {
        return if (fragment.arguments == null) {
            throw IllegalArgumentException("Fragment has no arguments.")
        } else if (Serializable::class.java.isAssignableFrom(mScreenClass)) {
            checkNotNull(
                fragment.requireArguments().getSerializable(KEY_SCREEN) as? ScreenT
            )
        } else if (Parcelable::class.java.isAssignableFrom(mScreenClass)) {
            checkNotNull(
                fragment.requireArguments().getParcelable<Parcelable>(KEY_SCREEN) as? ScreenT
            )
        } else {
            throw IllegalArgumentException("Screen " + mScreenClass.simpleName + " should be Serializable or Parcelable.")
        }
    }

    private fun checkNotNull(screen: ScreenT?): ScreenT {
        requireNotNull(screen) { "Failed to get screen from arguments of fragment." }
        return screen
    }

    companion object {
        private const val KEY_SCREEN = "me.aartikov.alligator.KEY_SCREEN"
    }
}